superstore/src/commonMain/kotlin/org/mobilenativefoundation/store/superstore5/impl/RealSuperstore.kt

```diff
@@ -0,0 +1,87 @@
+package org.mobilenativefoundation.store.superstore5.impl
+
+
+import kotlinx.coroutines.channels.ProducerScope
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.channelFlow
+import org.mobilenativefoundation.store.store5.Store
+import org.mobilenativefoundation.store.store5.StoreReadRequest
+import org.mobilenativefoundation.store.store5.StoreReadResponse
+import org.mobilenativefoundation.store.store5.StoreReadResponseOrigin
+import org.mobilenativefoundation.store.superstore5.Superstore
+import org.mobilenativefoundation.store.superstore5.SuperstoreResponse
+import org.mobilenativefoundation.store.superstore5.SuperstoreResponseOrigin
+import org.mobilenativefoundation.store.superstore5.Warehouse
+
+/**
+ * Coordinates [Store] and [Warehouse].
+ * Tries to fetch data through [Store].
+ * On failure, fetches data using [Warehouse] until a success.
+ */
+class RealSuperstore<Key : Any, Output : Any>(
+    private val store: Store<Key, Output>,
+    private val warehouses: List<Warehouse<Key, Output>>,
+) : Superstore<Key, Output> {
+    override fun get(key: Key, fresh: Boolean, refresh: Boolean): Flow<SuperstoreResponse<Output>> {
+        val request =
+            if (fresh) StoreReadRequest.fresh(key) else StoreReadRequest.cached(key, refresh)
+        return channelFlow {
+            store.stream(request).collect {
+                try {
+                    when (it) {
+                        is StoreReadResponse.Data -> useStore(
+                            it.requireData(),
+                            it.origin.toSuperstoreResponseOrigin()
+                        )
+
+                        is StoreReadResponse.Loading -> send(SuperstoreResponse.Loading)
+                        is StoreReadResponse.NoNewData -> {}
+                        is StoreReadResponse.Error.Exception -> useWarehouse(key)
+                        is StoreReadResponse.Error.Message -> useWarehouse(key)
+                    }
+                } catch (error: Throwable) {
+                    useWarehouse(key)
+                }
+            }
+        }
+    }
+
+    private suspend fun List<Warehouse<Key, Output>>.search(key: Key): Output {
+        for (warehouse in this) {
+            try {
+                val result = warehouse.get(key)
+                if (result != null) {
```

Comment:
For Warehouse, should we add some identifier, so the user can identify which warehouse the data is coming from? (For example they might be interesting to know if that is from a 2nd api or a hardcoded datasource, currently we can't not tell which warehouse is the data from)

Reply:
Hmm yeah I considered this but my thought was I wouldn't really care as a user from a UI POV where the data came from. But thinking again maybe we should at the very least to enable logging

Reply:
Modified `Warehouse` to have a name and `SuperstoreResponseOrigin.Warehouse` to have a value